pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -378,10 +384,170 @@ public CallableStatement prepareCall(String sql) throws SQLException {
 
   @Override
   public Map<String, Class<?>> getTypeMap() throws SQLException {
+    //#if mvn.project.property.postgresql.jdbc.spec < "JDBC4.1"
+    if (true) {
+      throw Driver.udtNotSupported(Connection.class, "getTypeMap()");
+    }
+    //#endif
     checkClosed();
+    // LinkedHashMap to maintain order of map provided by caller.
+    return new LinkedHashMap<String, Class<?>>(getTypeMapNoCopy());
```

Comment:
if the expectation is that the driver is maintaining a defensive copy, why do extra work to preserve order?

Reply:
Just to be nice to the API user.  So if they are debugging, what goes in matches what comes out later.  Makes things easier to read from debug log files, for example.

Reply:
so if a consumer passes in a `TreeMap` with a `String.CASE_INSENSITVE_ORDER` we should be concerned about preserving iteration order even though we do not care about changing the behavior of the map to operate on `hashCode` and `equals` rather than a provided comparator?

Absent using a `TreeMap`, it is unclear to me why consumers would care about iteration order.

Reply:
Most users wouldn't care.  But why not maintain the order?  No harm done.

Keep in mind, that the tutorials show to call getTypeMap(), make any changes, then call setTypeMap() again.  Now people can call setTypeMap() with any map they want, not necessarily the one they got from us.  But if they follow this common convention, they map will maintain the order they added the types.  This can be useful and comes at extremely little cost.

Ultimately, I'm doing this for ourselves.  We register quite a few custom types in our applications.  When I'm digging through log files debugging a problem, having consistent ordering of things helps, instead of stuff jumbled all around during a rehash.  This can make comparing different log lines via `diff --color` readily show the differences in the types mapped.  Sure, I could sort the maps when logging, but why not get this for free from the API?